### PR TITLE
graduate MatchLabelKeysInPodAffinity to Beta

### LINKED ISSUE #123156: PodAffinity/matchLabelKeys: Add tests to switch the feature gate
We should have a test for switching the feature gate, which is mentioned in the KEP but not implemented yet.
https://github.com/kubernetes/enhancements/pull/4450#discussion_r1478375329

We should make it in this release.

/sig scheduling
/kind feature
/assign

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

- graduate MatchLabelKeysInPodAffinity to Beta
- add the feature gate switching test.

#### Which issue(s) this PR fixes:

Fixes #123156
Part of https://github.com/kubernetes/enhancements/issues/3633

#### Does this PR introduce a user-facing change?

```release-note
Graduate MatchLabelKeys/MismatchLabelKeys feature in PodAffinity/PodAntiAffinity to Beta
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3633
```

### CODE REVIEW

**Comment by sanposhiho** [comment]
/retest 
https://github.com/kubernetes/kubernetes/issues/123621

File: test/integration/scheduler/filters/filters_test.go [hunk 1]
```diff
@@ -2074,3 +2075,142 @@ func TestUnschedulablePodBecomesSchedulable(t *testing.T) {
 		})
 	}
 }
+
+// TestPodAffinityMatchLabelKeyEnablement tests the Pod is correctly mutated by MatchLabelKeysInPodAffinity feature,
+// even if turing the feature gate enabled or disabled.
+func TestPodAffinityMatchLabelKeyEnablement(t *testing.T) {
```
Comment (sanposhiho) [inline test/integration/scheduler/filters/filters_test.go:2081]: The same is tested manually too, as suggested here: https://github.com/kubernetes/enhancements/pull/4450#discussion_r1480006648

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -3367,7 +3367,6 @@ type PodAffinityTerm struct {
 	// pod labels will be ignored. The default value is empty.
 	// The same key is forbidden to exist in both matchLabelKeys and labelSelector.
 	// Also, matchLabelKeys cannot be set when labelSelector isn't set.
-	// This is an alpha field and requires enabling MatchLabelKeysInPodAffinity feature gate.
```
Comment (alculquicondor) [inline staging/src/k8s.io/api/core/v1/types.go:3370]: This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).

Same below

**Review by alculquicondor — COMMENTED** [review]
/approve
/label api-review

**Comment by alculquicondor** [comment]
/assign @deads2k
for API, as alpha reviewer.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 2 above)
Comment (sanposhiho) [reply staging/src/k8s.io/api/core/v1/types.go:3370]: Updated.

**Review by alculquicondor — COMMENTED** [review]
/lgtm

**Comment by salehsedghpour** [comment]
Hi, Enhancements Team here, This PR needs `approve` label so that we can mark https://github.com/kubernetes/enhancements/issues/3633 as `tracked for code freeze`

**Comment by deads2k** [comment]
API change looks good.  Integration test looks good.  I strongly encourage writing a positive e2e test that could be (but doesn't necessarily need to be) made a conformance test in the future.

/approve

**Comment by alculquicondor** [comment]
@sanposhiho I guess we need to update the KEP for 1.31 now

**Comment by sanposhiho** [comment]
Made KEP update in https://github.com/kubernetes/enhancements/pull/4627.

@alculquicondor @deads2k 
I had to rebase the PR upon the master branch to fix the test failure. Please give it `lgtm` again when have time.

**Pull request merged**